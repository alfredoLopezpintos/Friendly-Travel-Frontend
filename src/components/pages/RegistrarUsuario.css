.form-box input[type="submit"] {
  border: 1px solid #242c37;
  border-radius: 5px;
  background-color: transparent;
}

.form-box input[type="submit"] {
  border: 1px solid #242c37;
  transition: all 0.3s ease-out;
  cursor: pointer;
  width: 100%;
}

.form-box input[type="submit"]:hover,
.form-box input[type="submit"]:hover {
  transition: all 0.3s ease-out;
  background: #242424;
  color: #fff;
  transition: 250ms;
}

/* CUSTOM CHECKBOX */
/* Customize the label (the container) */
.container {
  display: block;
  position: relative;
  padding-left: 35px;
  margin-bottom: 12px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

/* Hide the browser's default checkbox */
.container input {
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;
}

/* Create a custom checkbox */
.checkmark {
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
}

/* On mouse-over, add a grey background color */
.container:hover input ~ .checkmark {
  background-color: #ccc;
}

/* When the checkbox is checked, add a blue background */
.container input:checked ~ .checkmark {
  background-color: #2196F3;
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark:after {
  content: "";
  position: absolute;
  display: none;
}

/* Show the checkmark when checked */
.container input:checked ~ .checkmark:after {
  display: block;
}

/* Style the checkmark/indicator */
.container .checkmark:after {
  left: 9px;
  top: 5px;
  width: 5px;
  height: 10px;
  border: solid white;
  border-width: 0 3px 3px 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

#checkBox {
  font-size: 12px;
  padding-bottom: 10px;
}
