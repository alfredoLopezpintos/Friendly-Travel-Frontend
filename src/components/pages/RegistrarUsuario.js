import axios from "axios";
import es from "date-fns/locale/es";
import moment from "moment";
import React from "react";
import "react-datepicker/dist/react-datepicker.css";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import configData from "../../configData.json";
import "./Login.css";
import { toast } from "react-toastify";
import "./RegistrarUsuario.css";
import { registerLocale } from "react-datepicker";
import { transformDate2 } from "../Utilities";
import {
  isValidDocument,
  isValidEmail,
  isValidPhoneNumber
} from "../../utils/ValidationFunctions";
registerLocale("es", es);

export default function RegistrarUsuario() {
  const { register, handleSubmit } = useForm();
  const onSubmit = (data, e) => postData(data, e);
  const onError = (errors, e) => console.log(errors, e);
  const history = useHistory();
  let checkBox = false;

  function borrarCampos(data) {
    checkBox = false;
  }

  function formValidate(data) {
    if (
      data.email === "" ||
      data.name === "" ||
      data.surname === "" ||
      data.birthDate === "" ||
      data.documentId === "" ||
      data.phoneNumber === ""
    ) {
      toast.error("Debe completar todos los campos");
      return false;
    } else if (!moment(data.birthDate, "DD-MM-YYYY").isValid()) {
      toast.error("Fecha inválida");
      return false;
    } else if (moment().diff(data.birthDate, "years") < 18) {
      console.log(moment().diff(data.birthDate, "years") < 18);
      toast.error("El usuario debe ser mayor de edad");
      return false;
    } else if (!isValidEmail(data.email)) {
      toast.error("El formato del correo electrónico no es válido");
      return false;
    } else if (!isValidPhoneNumber(data.phoneNumber)) {
      toast.error("El formato del teléfono no es válido");
      return false;
    } else if (!isValidDocument(data.documentId)) {
      return false;
    } else if (!checkBox) {
      toast.error(
        "Debe estar de acuerdo con la política de uso de FriendlyTravel" +
        " para poder registrarse."
      );
      return false;
    } else {
      return true;
    }
  }

  async function postData(data, e) {
    console.log(data);
    if (formValidate(data)) {
      data.birthDate = transformDate2(data.birthDate);
      const postUserEndpoint = configData.AWS_REST_ENDPOINT + "/users";

      toast.promise(
        axios
          .post(postUserEndpoint, data)
          .then((response) => {
            redirect();
          })
          .catch((error) => {
            console.error(error);
          }),
        {
          pending: {
            render() {
              return "Cargando";
            },
            icon: true,
          },
          error: {
            render({ data }) {
              toast.error(data.response.data.message);
            },
          },
        }
      );
    }
  }

  const handleCheckBoxChange = (event) => {
    if (event.target.checked) {
      checkBox = true;
    } else {
      checkBox = false;
    }
  };

  async function redirect(data, e) {
    toast.success("Usuario creado correctamente.");
    history.push("/");
    borrarCampos(data);
  }

  return (
    <>
      <div>
        <div className="grid align__item">
          <div className="register">
            <div className="big_logo">
              <img
                src={require("../../assets/images/logo2.png")}
                alt="travel logo"
                width={200}
              ></img>
            </div>
            <br />
            <h2> Registrar usuario </h2>
            <br />
            <form onSubmit={handleSubmit(onSubmit, onError)} className="form" data-testid="form">
              <div>
                <label>Nombre</label>
                <div className="form__field">
                  <input
                    {...register("name")}
                    placeholder="Ingrese su nombre"
                    type="text"
                  />
                </div>

                <label>Apellido</label>
                <div className="form__field">
                  <input
                    {...register("surname")}
                    placeholder="Ingrese su apellido"
                    type="text"
                  />
                </div>

                <label>Email</label>
                <div className="form__field">
                  <input
                    {...register("email")}
                    placeholder="Ingrese su email"
                    type="email"
                  />
                </div>

                <label>Fecha de nacimiento</label>
                <div className="form__field" data-testid="birthDate" >
                  <input
                    {...register("birthDate")}
                    type="date"
                    format="dd-MM-yyyy"
                  />
                </div>

                <label>Cédula de identidad</label>
                <div className="form__field">
                  <input
                    {...register("documentId")}
                    placeholder="Sin puntos ni guiones, ej. 43215678"
                    type="number"
                  />
                </div>

                <label>Teléfono de contacto</label>
                <div className="form__field">
                  <input
                    {...register("phoneNumber")}
                    placeholder="Número de teléfono, ej. 099111333"
                    type="number"
                  />
                </div>
                <div className="form__field">
                  <label id="checkBox" className="container">
                    Confirmo haber leído y estar de acuerdo con las
                    <a href="/policy"> políticas de uso de FriendlyTravel</a>
                    <input type="checkbox" onChange={handleCheckBoxChange} />
                    <span className="checkmark"></span>
                  </label>
                </div>
                <br />
                <div className="form__field">
                  <input type="submit" value="Aceptar" />
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </>
  );
}