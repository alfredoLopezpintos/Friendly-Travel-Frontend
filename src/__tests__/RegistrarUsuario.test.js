import axios from 'axios';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import RegistrarUsuario from '../components/pages/RegistrarUsuario';

jest.mock('axios');

describe('RegistrarUsuario', () => {
  beforeEach(() => {
    render(<MemoryRouter><RegistrarUsuario /></MemoryRouter>);
  });

  it('renders RegistrarUsuario component', () => {
    const formElement = screen.getByTestId('form');
    expect(formElement).toMatchSnapshot();
  });

  it('form validation works - without filling out any fields', () => {
    const nameInput = screen.getByPlaceholderText('Ingrese su nombre');
    const surnameInput = screen.getByPlaceholderText('Ingrese su apellido');
    const emailInput = screen.getByPlaceholderText('Ingrese su email');
    const birthdateInput = screen.getByTestId('birthDate');
    const inputElement = birthdateInput.querySelector('input')
    const documentIdInput = screen.getByPlaceholderText('Sin puntos ni guiones, ej. 43215678');
    const phoneNumberInput = screen.getByPlaceholderText('Número de teléfono, ej. 099111333');
    const checkboxElement = screen.getByLabelText('Confirmo haber leído y estar de acuerdo con las políticas de uso de FriendlyTravel');
    const formElement = screen.getByTestId('form');

    fireEvent.submit(formElement);
    (() =>
      expect(getByText("Debe completar todos los campos")).toBeInTheDocument()
    );
  });

  it('form validation works - Fill out fields incorrectly', () => {
    const nameInput = screen.getByPlaceholderText('Ingrese su nombre');
    const surnameInput = screen.getByPlaceholderText('Ingrese su apellido');
    const emailInput = screen.getByPlaceholderText('Ingrese su email');
    const birthdateInput = screen.getByTestId('birthDate');
    const inputElement = birthdateInput.querySelector('input')
    const documentIdInput = screen.getByPlaceholderText('Sin puntos ni guiones, ej. 43215678');
    const phoneNumberInput = screen.getByPlaceholderText('Número de teléfono, ej. 099111333');
    const checkboxElement = screen.getByLabelText('Confirmo haber leído y estar de acuerdo con las políticas de uso de FriendlyTravel');
    const formElement = screen.getByTestId('form');

    fireEvent.change(nameInput, { target: { value: '' } });
    fireEvent.change(surnameInput, { target: { value: '' } });
    fireEvent.change(emailInput, { target: { value: 'invalidEmail.com' } });
    fireEvent.change(inputElement, { target: { value: 'invalidDate' } });
    fireEvent.change(documentIdInput, { target: { value: '123' } });
    fireEvent.change(phoneNumberInput, { target: { value: '123' } });
    fireEvent.click(checkboxElement);
    fireEvent.submit(formElement);
    (() =>
      expect(getByText("El formato del correo electrónico no es válido")).toBeInTheDocument()
    );
    (() =>
      expect(getByText("Fecha inválida")).toBeInTheDocument()
    );
    (() =>
      expect(getByText("Debe estar de acuerdo con las políticas de uso de FriendlyTravel para poder registrarse.")).toBeInTheDocument()
    );
  });

  it('form validation works - Fill out fields correctly', () => {
    const nameInput = screen.getByPlaceholderText('Ingrese su nombre');
    const surnameInput = screen.getByPlaceholderText('Ingrese su apellido');
    const emailInput = screen.getByPlaceholderText('Ingrese su email');
    const birthdateInput = screen.getByTestId('birthDate');
    const inputElement = birthdateInput.querySelector('input')
    const documentIdInput = screen.getByPlaceholderText('Sin puntos ni guiones, ej. 43215678');
    const phoneNumberInput = screen.getByPlaceholderText('Número de teléfono, ej. 099111333');
    const checkboxElement = screen.getByLabelText('Confirmo haber leído y estar de acuerdo con las políticas de uso de FriendlyTravel');
    const formElement = screen.getByTestId('form');

    fireEvent.change(nameInput, { target: { value: 'Test' } });
    fireEvent.change(surnameInput, { target: { value: 'User' } });
    fireEvent.change(emailInput, { target: { value: 'testuser@mail.com' } });
    fireEvent.change(inputElement, { target: { value: '11-11-2000' } });
    fireEvent.change(documentIdInput, { target: { value: '12345678' } });
    fireEvent.change(phoneNumberInput, { target: { value: '099111333' } });
    fireEvent.click(checkboxElement);
    fireEvent.submit(formElement);
    (() =>
      expect(axios.post).toHaveBeenCalled()
    );
  });
});
