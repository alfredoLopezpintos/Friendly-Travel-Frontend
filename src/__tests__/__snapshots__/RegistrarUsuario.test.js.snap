// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrarUsuario renders RegistrarUsuario component 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div>
    <label>
      Nombre
    </label>
    <div
      class="form__field"
    >
      <input
        name="name"
        placeholder="Ingrese su nombre"
        type="text"
      />
    </div>
    <label>
      Apellido
    </label>
    <div
      class="form__field"
    >
      <input
        name="surname"
        placeholder="Ingrese su apellido"
        type="text"
      />
    </div>
    <label>
      Email
    </label>
    <div
      class="form__field"
    >
      <input
        name="email"
        placeholder="Ingrese su email"
        type="email"
      />
    </div>
    <label>
      Fecha de nacimiento
    </label>
    <div
      class="form__field"
      data-testid="birthDate"
    >
      <input
        format="dd-MM-yyyy"
        name="birthDate"
        type="date"
      />
    </div>
    <label>
      Cédula de identidad
    </label>
    <div
      class="form__field"
    >
      <input
        name="documentId"
        placeholder="Sin puntos ni guiones, ej. 43215678"
        type="number"
      />
    </div>
    <label>
      Teléfono de contacto
    </label>
    <div
      class="form__field"
    >
      <input
        name="phoneNumber"
        placeholder="Número de teléfono, ej. 099111333"
        type="number"
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="container"
        id="checkBox"
      >
        Confirmo haber leído y estar de acuerdo con las
        <a
          href="/policy"
        >
           políticas de uso de FriendlyTravel
        </a>
        <input
          type="checkbox"
        />
        <span
          class="checkmark"
        />
      </label>
    </div>
    <br />
    <div
      class="form__field"
    >
      <input
        type="submit"
        value="Aceptar"
      />
    </div>
  </div>
</form>
`;
